### PR TITLE
drivers: can: Add frontend for Atmel SAM

### LINKED ISSUE #24802: Implement Microchip SAM family CAN controller driver
### Summary

A number of Microchip (formerly, Atmel) SAM family SoCs include a CAN 2.0A/B and CAN-FD capable host controller peripheral.

In order to take advantage of this, it is necessary to implement a driver for it, that can interface to the existing Zephyr CAN protocol stack.

### Target Devices

The following Microchip SAM family devices implement the "Microchip CAN Controller" peripheral:
* SAM C21
* SAM E51/E54
* SAM E70
* SAM V70/V71

Note that the following (old generation) Atmel SAM family devices also include a CAN controller peripheral, but this is completely different from the "Microchip CAN Controller" and therefore incompatible; this issue does not propose to support them:
* SAM 3X
* SAM 4E

### PR DESCRIPTION
**Pull request by alexanderwachter**

### CODE REVIEW

**Comment by alexanderwachter** [comment]
ping @JeremyBCD

**Comment by alexanderwachter** [comment]
Rebased and fixed issue in can_mcan.c (drivers: can: mcan: fix array index in chache handling param )

File: boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig [hunk 1]
```diff
@@ -11,6 +11,7 @@ CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SERIAL=y
 CONFIG_USART_SAM=y
+CONFIG_GPIO=y
```
Comment (nandojve) [inline boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig:14]: #41550 should solve GPIO dependency.

File: drivers/can/Kconfig.sam [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# SAM CAN configuration options
+# Copyright (c) 2021 Alexander Wachter
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_ATMEL_SAM_CAN := atmel,sam-can
+
+config CAN_SAM
+	bool "Atmel SAM CAN driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_ATMEL_SAM_CAN))
+	select CAN_MCAN
```
Comment (nandojve) [inline drivers/can/Kconfig.sam:10]: Shouldn't CAN_SAM have a depends on CAN too?

File: drivers/can/can_sam.c [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+#include <kernel.h>
+#include "can_sam.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+#define DT_DRV_COMPAT atmel_sam_can
+
+int can_sam_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+
```
Comment (nandojve) [inline drivers/can/can_sam.c:18]: Empty line.

File: dts/bindings/can/atmel,sam-can.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      required: true
+      description: peripheral ID
+    pinctrl-0:
+      type: phandles
+      required: true
+      description: |
+        PIO configuration for CAN_RX and CAN_TX. We expect that the phandles will
+        reference pinctrl nodes.  These nodes will have a nodelabel that
+        matches the Atmel SoC HAL defines and be of the form
+          pinctrl-0 = <&pa8a_uart_urxd &pa9a_uart_utxd>;
```
Comment (nandojve) [inline dts/bindings/can/atmel,sam-can.yaml:19]: This example should use can pins instead of uart.

**Review by nandojve — CHANGES_REQUESTED** [review]
Hello @alexanderwachter ,

Looks good, just few.

File: boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi [hunk 5]
```diff
@@ -150,3 +151,14 @@ zephyr_udc0: &usbhs {
...
+&can0 {
+	status = "okay";
+	pinctrl-0 = <&pb3a_can0_rx0 &pb2a_can0_tx0>;
+	bus-speed = <125000>;
+	sjw = <1>;
+	sample-point = <875>;
+	bus-speed-data = <1000000>;
+	sjw-data = <1>;
+	sample-point-data = <875>;
+};
```
Comment (henrikbrixandersen) [inline boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi:164]: I suggest moving `bus-speed`, `sjw`, `bus-speed-data`, and `sjw-data` to SoC DTSI since these will likely be identical for most boards.

File: drivers/can/Kconfig.sam (hunk 2 above)
Comment (henrikbrixandersen) [reply drivers/can/Kconfig.sam:10]: No, it's already wrapped in `if CAN` in the higher level Kconfig file.

File: drivers/can/Kconfig.sam [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+if CAN_SAM
+
+config CAN_SAM_CKDIV
+	int "Clock divider"
+	range 0 255
+	default 0
+	depends on CAN_SAM
+	help
+	  Clock divider for the MCAN core clock.
```
Comment (henrikbrixandersen) [inline drivers/can/Kconfig.sam:20]: Shouldn't this be configured via DTS instead?

File: drivers/can/can_sam.c [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+	*rate = SOC_ATMEL_SAM_MCK_FREQ_HZ / (CONFIG_CAN_SAM_CKDIV + 1);
+	return 0;
+}
+
+void can_sam_set_state_change_callback(const struct device *dev,
+				       can_state_change_callback_t isr)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:24]: This needs rebasing.

File: drivers/can/can_sam.c [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+{
+	struct can_sam_data *data = DEV_DATA(dev);
+
+	data->mcan_data.state_change_cb = isr;
+}
+
+void can_sam_clock_enable(const struct can_sam_config *cfg)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:31]: Shouldn't all of these functions be `static`?

File: drivers/can/can_sam.c [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+	REG_PMC_PCK5 = PMC_PCK_CSS_PLLA_CLK | PMC_PCK_PRES(CONFIG_CAN_SAM_CKDIV);
+	PMC->PMC_SCER |= PMC_SCER_PCK5;
+	soc_pmc_peripheral_enable(cfg->pmc_id);
+}
+
+int can_sam_init(const struct device *dev)
+{
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_sam_data *data = DEV_DATA(dev);
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:42]: Please, do not use `DEV_DATA()`/`DEV_CFG()` in new code. See #37616 for details.

File: drivers/can/can_sam.c [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+{
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+
+	return can_mcan_get_state(mcan_cfg, err_cnt);
+}
+
+int can_sam_send(const struct device *dev, const struct zcan_frame *frame,
+		  k_timeout_t timeout, can_tx_callback_t callback,
+		  void *callback_arg)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:70]:
```suggestion
		  void *user_data)
```

File: drivers/can/can_sam.c [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+			  void *cb_arg, const struct zcan_filter *filter)
+{
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	return can_mcan_add_rx_filter(mcan_data, msg_ram, isr, cb_arg, filter);
+}
+
+void can_sam_remove_rx_filter(const struct device *dev, int filter_nr)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:92]:
```suggestion
void can_sam_remove_rx_filter(const struct device *dev, int filter_id)
```

File: drivers/can/can_sam.c [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	return can_mcan_send(mcan_cfg, mcan_data, msg_ram, frame, timeout,
+			     callback, callback_arg);
+}
+
+int can_sam_add_rx_filter(const struct device *dev, can_rx_callback_t isr,
+			  void *cb_arg, const struct zcan_filter *filter)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:83]:
```suggestion
int can_sam_add_rx_filter(const struct device *dev, can_rx_callback_t callback,
			  void *user_data, const struct zcan_filter *filter)
```

File: drivers/can/can_sam.c [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+		       const struct can_timing *timing,
+		       const struct can_timing *timing_data)
+{
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+
+	return can_mcan_set_timing(mcan_cfg, timing, timing_data);
+}
+
+void can_sam_line_0_isr(void *arg)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:119]:
```suggestion
void can_sam_line_0_isr(const struct device *dev)
```

File: drivers/can/can_sam.c [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+	const struct can_sam_config *cfg = DEV_CFG(dev);
+	const struct can_mcan_config *mcan_cfg = &cfg->mcan_cfg;
+	struct can_sam_data *data = DEV_DATA(dev);
+	struct can_mcan_data *mcan_data = &data->mcan_data;
+	struct can_mcan_msg_sram *msg_ram = &data->msg_ram;
+
+	can_mcan_line_0_isr(mcan_cfg, msg_ram, mcan_data);
+}
+
+void can_sam_line_1_isr(void *arg)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:131]:
```suggestion
void can_sam_line_1_isr(const struct device *dev)
```

File: drivers/can/can_sam.c [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+		.sjw = 0x7f,
+		.prop_seg = 0x00,
+		.phase_seg1 = 0x100,
+		.phase_seg2 = 0x80,
+		.prescaler = 0x200
+	},
+#ifdef CONFIG_CAN_FD_MODE
+	.timing_min_data = {
+		.sjw = 0x01,
+		.prop_seg = 0x01,
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:172]: Shouldn't this be 0x00? How come its min value is higher than it's max value?

File: drivers/can/can_sam.c [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+DEVICE_DT_INST_DEFINE(inst, &can_sam_init, NULL,                               \
+		      &can_sam_dev_data_##inst, &can_sam_cfg_##inst,           \
+		      POST_KERNEL, CONFIG_CAN_INIT_PRIORITY,                   \
+		      &can_api_funcs);
+
+#define CAN_SAM_INST(inst)                                                     \
+CAN_SAM_IRQ_CFG_FUNCTION(inst)                                                 \
+CAN_SAM_CFG_INST(inst)                                                         \
+CAN_SAM_DATA_INST(inst)                                                        \
+CAN_SAM_DEVICE_INST(inst)
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.c:258]: Lots of macros going on here, some just for defining a single line. I would collapse as much as this into one macro as we tend to do elsewhere in Zephyr (see e.g. the NXP MCUX M_CAN frontend driver).

File: drivers/can/can_sam.h [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+	struct soc_gpio_pin pin_list[2];
+	uint8_t pmc_id;
+};
+
+struct can_sam_data {
+	struct can_mcan_data mcan_data;
+	struct can_mcan_msg_sram msg_ram;
+};
+
+#endif /* ZEPHYR_DRIVERS_CAN_SAM_H_ */
```
Comment (henrikbrixandersen) [inline drivers/can/can_sam.h:29]: There's not much point of adding a such a small header that is only used in one implementation file. Please just put the structs at the top of the C-file.

File: samples/net/sockets/can/prj.conf [hunk 18]
```diff
@@ -1,4 +1,5 @@
 CONFIG_CAN=y
+CONFIG_NET_L2_ETHERNET=n
```
Comment (henrikbrixandersen) [inline samples/net/sockets/can/prj.conf:2]: Shouldn't this be a separate PR?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: dts/bindings/can/atmel,sam-can.yaml (hunk 4 above)
Comment (alexanderwachter) [reply dts/bindings/can/atmel,sam-can.yaml:19]: Whoops. cpy&paste :-)

File: boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi (hunk 5 above)
Comment (alexanderwachter) [reply boards/arm/sam_e70_xplained/sam_e70_xplained-common.dtsi:164]: Yes, but these are typically board-specific configurations. I'll keep them the same as the stm32fd.

File: drivers/can/Kconfig.sam (hunk 6 above)
Comment (alexanderwachter) [reply drivers/can/Kconfig.sam:20]: Not sure. This config applies to each controller.
For stm32 it is also a Kconfig

File: drivers/can/can_sam.h (hunk 17 above)
Comment (alexanderwachter) [reply drivers/can/can_sam.h:29]: I don't agree generally, but in this case, yes.

File: samples/net/sockets/can/prj.conf (hunk 18 above)
Comment (alexanderwachter) [reply samples/net/sockets/can/prj.conf:2]: No. It's a separate commit.

File: drivers/can/can_sam.c [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+{
+
+	*rate = SOC_ATMEL_SAM_MCK_FREQ_HZ / (CONFIG_CAN_SAM_CKDIV + 1);
+	return 0;
+}
+
+void can_sam_set_state_change_callback(const struct device *dev,
+				       can_state_change_callback_t isr)
+{
+	struct can_sam_data *data = DEV_DATA(dev);
```
Comment (gmarull) [inline drivers/can/can_sam.c:26]: Please, don't use `DEV_DATA/DEV_CONFIG` macros in new code, ref. https://github.com/zephyrproject-rtos/zephyr/issues/37616

**Review by gmarull — DISMISSED** [review]

File: drivers/can/can_sam.c (hunk 19 above)
Comment (alexanderwachter) [reply drivers/can/can_sam.c]: Done.

File: drivers/can/can_sam.c (hunk 16 above)
Comment (alexanderwachter) [reply drivers/can/can_sam.c]: I think it is much more readable and easier to debug this way.
Most of the macros are more than one line. To be consistent, I added macros for all definitions.

File: boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig (hunk 1 above)
Comment (alexanderwachter) [reply boards/arm/sam_e70_xplained/sam_e70_xplained_defconfig:14]: I'll remove this line and rebase when your PR is merged

**Review by henrikbrixandersen — APPROVED** [review]

**Review by nandojve — APPROVED** [review]
